scipy/_lib/_disjoint_set.py

```diff
@@ -0,0 +1,171 @@
+"""
+Disjoint set data structure
+"""
+import collections
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_subsets : int
+        The number of subsets.
+
+    Methods
+    -------
+    add
+    merge
+    connected
+    __getitem__
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    or *merge-find* data structure. The *find* operation (implemented in
+    `__getitem__`)* implements the *path halving* variant. The *merge* method
+    implements the *merge by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    Examples
+    --------
+    >>> from scipy.cluster.hierarchy import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> disjoint_set = DisjointSet([1, 2, 3, 'a', 'b'])
+
+    Merge some subsets:
+
+    >>> disjoint_set.merge(1, 2)
+    True
+    >>> disjoint_set.merge(3, 'a')
+    True
+    >>> disjoint_set.merge('a', 'b')
+    True
+    >>> disjoint_set.merge('b', 'b')
+    False
+
+    Find root elements:
+
+    >>> disjoint_set[2]
+    1
+    >>> disjoint_set['b']
+    3
+
+    Test connectivity:
+
+    >>> disjoint_set.connected(1, 2)
+    True
+    >>> disjoint_set.connected(1, 'b')
+    False
+
+    List elements in disjoint set:
+
+    >>> list(disjoint_set)
+    [1, 2, 3, 'a', 'b']
+    """
+    def __init__(self, elements=None):
+        self.n_subsets = 0
+        self._sizes = {}
+        self._parents = {}
+        # _indices tracks the element insertion order
```

Comment:
Obvious as it may be, it would have helped me ever so slightly if it was spelled out that the `OrderedDict` is necessary only for `iter(self._indices)` below.

Reply:
OK - I have added this comment:
```
        # _indices tracks the element insertion order - OrderedDict is used to
        # ensure correct ordering in `__iter__` and for tie-breaking in
        # `merge`.
```

Reply:
A bit nitpicky I know, but from what I can tell, the fact that it's an `OrderedDict` is not actually used during tie-breaking: There, all your using is that `self._indices[x]` counts the insertion order, which you would also achieve with an ordinary `dict`.

Reply:
I went with an `OrderedDict` because this behaviour has only been "official" since python 3.7. From the release notes:

> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was implementation detail of CPython from 3.6.

Recommendations on the internet suggest using `OrderedDict` for python 3.6 because other python implementations might not have this behaviour, and SciPy has not yet dropped support for 3.6.

Reply:
Does that make sense?

Reply:
I assume you mean the part of `merge` relevant to tie-breaking defined in the conditional

```python
if (sizes[xr], self._indices[yr]) < (sizes[yr], self._indices[xr]):
```

right? If so, that would work just the same if `self._indices` were a `dict`, even prior to CPython's adoption of the ordered dict implementation; all that matters is the value of `self._indices[xr]` resp. `...[yr]`, and you're setting that explicitly.

Reply:
Oh, right - I'm on the same page now. You are absolutely correct. I have fixed the comment.